crucible/src/Lang/Crucible/CFG/Generator.hs

```diff
@@ -480,58 +583,69 @@ endCurrentBlock term = End $ do
   -- Store block
   seq b $ do
   gsBlocks %= (Seq.|> b)
-
--- | Resume execution by jumping to a label.
-resume_  :: Label s -- ^ Label to jump to.
-         -> (() -> End ext h s t ret ())
-            -- ^ Continuation to run.
-         -> End ext h s t ret ()
-resume_ lbl c = End $ do
-  checkCurrentUnassigned
-  gsCurrent .= Just (initCurrentBlockState Set.empty (LabelID lbl))
-  unEnd $ c ()
-
--- | Resume execution by jumping to a lambda label.
-resume :: LambdaLabel s r
-          -- ^ Label to jump to.
-       -> (Expr ext s r -> End ext h s t ret ())
-          -- ^ Continuation to run.
-       -> End ext h s t ret ()
-resume lbl c = End $ do
-  let block_id = LambdaID lbl
-  checkCurrentUnassigned
-  gsCurrent .= Just (initCurrentBlockState Set.empty block_id)
-  unEnd $ c (AtomExpr (lambdaAtom lbl))
-
-defineSomeBlock :: BlockID s
-                -> Generator ext h s t ret ()
-                -> End ext h s t ret ()
-defineSomeBlock l next = End $ do
-  gs <- get
-  let gs_next = gs & gsCurrent .~ Just (initCurrentBlockState Set.empty l)
-  let c_next _ gs' =
-        let p' = gs'^.gsPosition
-         in error $ "Block at " ++ show p' ++ " ended without terminating."
-  gs' <- liftST $ runStateContT (unGenerator next) c_next gs_next
+-}
+
+-- | End the translation of the current block, and then start a new
+-- block with the given label.
+resume_ ::
+  IsSyntaxExtension ext =>
+  TermStmt s ret ->
+  End ext h s t ret (Label s) ->
+  Generator ext h s t ret ()
+resume_ term m =
+  Generator $
+  bindIxGenerator (terminateBlock term) $ \ () ->
+  bindIxGenerator (unEnd m) $ \ lbl ->
+  startBlock (LabelID lbl)
+
+-- | End the translation of the current block, and then start a new
+-- lambda block with the given label.
+resume ::
+  IsSyntaxExtension ext =>
+  TermStmt s ret ->
+  End ext h s t ret (LambdaLabel s r) ->
+  Generator ext h s t ret (Expr ext s r)
+resume term m =
+  Generator $
+  bindIxGenerator (terminateBlock term) $ \ () ->
+  bindIxGenerator (unEnd m) $ \ lbl ->
+  bindIxGenerator (startBlock (LambdaID lbl)) $ \ () ->
+  pureIxGenerator (AtomExpr (lambdaAtom lbl))
+
+defineSomeBlock ::
+  IsSyntaxExtension ext =>
+  BlockID s ->
+  Generator ext h s t ret (TermStmt s ret) ->
+  End ext h s t ret ()
+defineSomeBlock l next =
+  End $
+  bindIxGenerator getIxGenerator $ \gs ->
+  bindIxGenerator (startBlock l) $ \() ->
+  bindIxGenerator (unGenerator next) $ \term ->
+  bindIxGenerator (terminateBlock term) $ \() ->
+  bindIxGenerator getIxGenerator $ \gs' ->
   -- Reset current block and state.
   put $ gs' & gsCurrent  .~ gs^.gsCurrent
             & gsPosition .~ gs^.gsPosition
             & gsState    .~ gs^.gsState
```

Comment:
I'm a bit surprised to see `gsState` getting restored here.  Is this actually the desired behavior?